File: ios/ProviiWallet/ProviiWallet/Features/Search/SearchManager.swift
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust

import SwiftUI
import Combine

// In-app search manager maintaining an index of searchable screens, settings, help
// topics, and features. Performs case-insensitive filtering across titles, keywords,
// and descriptions with debounced query updates to keep the UI responsive.

// MARK: - Searchable Item

enum SearchableItemType {
    case screen
    case setting
    case helpTopic
    case feature
}

struct SearchableItem: Identifiable, Hashable {
    let id = UUID()
    let title: String
    let subtitle: String?
    let keywords: [String]
    let type: SearchableItemType
    let destination: SearchDestination
    let icon: String
    let iconColor: Color

    func matches(query: String) -> Bool {
        let lowercasedQuery = query.lowercased()
        return title.lowercased().contains(lowercasedQuery) ||
               (subtitle?.lowercased().contains(lowercasedQuery) ?? false) ||
               keywords.contains { $0.lowercased().contains(lowercasedQuery) }
    }

    static func == (lhs: SearchableItem, rhs: SearchableItem) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

enum SearchDestination {
    case accessibilitySettings
    case settings
    case credentials
    case help
    case whereToGet
    case languageSelection
    case specificHelpTopic(HelpTopic)
}

// MARK: - Search Manager

@MainActor
class SearchManager: ObservableObject {
    @Published var searchQuery = ""
    @Published var searchResults: [SearchableItem] = []
    @Published var isSearching = false

    static let shared = SearchManager()

    private let allSearchableItems: [SearchableItem]

    private init() {
        // Initialize all searchable items
        self.allSearchableItems = SearchManager.createSearchableItems()
    }

    func search(query: String) {
        searchQuery = query

        guard !query.isEmpty else {
            searchResults = []
            isSearching = false
            return
        }

        isSearching = true

        // Filter items that match the query
        let matches = allSearchableItems.filter { $0.matches(query: query) }

        // Sort by relevance (title matches first, then subtitle, then keywords)
        searchResults = matches.sorted { item1, item2 in
            let query = query.lowercased()

            // Exact matches first
            if item1.title.lowercased() == query && item2.title.lowercased() != query {
                return true
            }
            if item2.title.lowercased() == query && item1.title.lowercased() != query {
                return false
            }

            // Title starts with query
            if item1.title.lowercased().hasPrefix(query) && !item2.title.lowercased().hasPrefix(query) {
                return true
            }
            if item2.title.lowercased().hasPrefix(query) && !item1.title.lowercased().hasPrefix(query) {
                return false
            }

            // Title contains query
            if item1.title.lowercased().contains(query) && !item2.title.lowercased().contains(query) {
                return true
            }
            if item2.title.lowercased().contains(query) && !item1.title.lowercased().contains(query) {
                return false
            }

            // Alphabetical order
            return item1.title < item2.title
        }
    }

    func clearSearch() {
        searchQuery = ""
        searchResults = []
        isSearching = false
    }

    private static func createSearchableItems() -> [SearchableItem] {
        var items: [SearchableItem] = []

        // Accessibility Settings (HIGH PRIORITY)
        items.append(SearchableItem(
            title: LocalizedString.searchAccessibilitySettings.localized,
            subtitle: LocalizedString.searchCustomizeExperience.localized,
            keywords: ["accessibility", "a11y", "settings", "customize", "screen reader", "voiceover", "talkback", "large text", "high contrast", "voice", "speech", "disabilities", "wcag", "ada"],
            type: .screen,
            destination: .accessibilitySettings,
            icon: "accessibility",
            iconColor: .blue
        ))

        // Individual Accessibility Features
        items.append(SearchableItem(
            title: LocalizedString.searchLargeText.localized,
            subtitle: LocalizedString.searchIncreaseTextSize.localized,
            keywords: ["large", "text", "size", "font", "bigger", "accessibility", "vision", "sight"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "textformat.size",
            iconColor: .blue
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchHighContrast.localized,
            subtitle: LocalizedString.searchEnhanceVisibility.localized,
            keywords: ["high", "contrast", "visibility", "vision", "accessibility", "colors", "see better"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "circle.lefthalf.filled",
            iconColor: .blue
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchVoiceInput.localized,
            subtitle: LocalizedString.searchControlWithVoice.localized,
            keywords: ["voice", "speech", "input", "control", "speak", "talk", "accessibility", "hands-free"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "mic.fill",
            iconColor: .blue
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchManualCodeEntry.localized,
            subtitle: LocalizedString.searchTypeCodesInsteadScanning.localized,
            keywords: ["manual", "code", "entry", "type", "keyboard", "accessibility", "qr", "scan"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "keyboard",
            iconColor: .blue
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchSimplifiedUI.localized,
            subtitle: LocalizedString.searchReduceVisualComplexity.localized,
            keywords: ["simplified", "simple", "ui", "interface", "easy", "basic", "accessibility", "cognitive"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "square.grid.2x2",
            iconColor: .blue
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchColorBlindnessSupport.localized,
            subtitle: LocalizedString.searchAdjustColorsVisibility.localized,
            keywords: ["color", "blind", "blindness", "vision", "accessibility", "deuteranopia", "protanopia", "tritanopia"],
            type: .setting,
            destination: .accessibilitySettings,
            icon: "eyedropper",
            iconColor: .blue
        ))

        // Settings
        items.append(SearchableItem(
            title: LocalizedString.searchSettings.localized,
            subtitle: LocalizedString.searchAppConfiguration.localized,
            keywords: ["settings", "config", "configuration", "preferences", "options"],
            type: .screen,
            destination: .settings,
            icon: "gearshape.fill",
            iconColor: AccessibleColors.secondaryText
        ))

        if LanguageSettings.LanguageInfo.hasMultipleLanguages {
            items.append(SearchableItem(
                title: LocalizedString.searchLanguage.localized,
                subtitle: LocalizedString.searchChangeAppLanguage.localized,
                keywords: ["language", "translate", "locale", "español", "français", "deutsch", "italiano", "português"],
                type: .screen,
                destination: .languageSelection,
                icon: "globe",
                iconColor: .green
            ))
        }

        // Credentials
        items.append(SearchableItem(
            title: LocalizedString.searchMyCredentials.localized,
            subtitle: LocalizedString.searchViewCredentials.localized,
            keywords: ["credentials", "wallet", "id", "verification", "identity"],
            type: .screen,
            destination: .credentials,
            icon: "wallet.pass.fill",
            iconColor: .purple
        ))

        items.append(SearchableItem(
            title: LocalizedString.searchGetCredential.localized,
            subtitle: LocalizedString.searchFindIssuers.localized,
            keywords: ["get", "obtain", "acquire", "credential", "issuer", "location", "where"],
            type: .screen,
            destination: .whereToGet,
            icon: "plus.circle.fill",
            iconColor: .blue
        ))

        // Help Topics
        items.append(SearchableItem(
            title: LocalizedString.searchHelp.localized,
            subtitle: LocalizedString.searchGetAssistance.localized,
            keywords: ["help", "support", "assistance", "guide", "tutorial", "how to"],
            type: .screen,
            destination: .help,
            icon: "questionmark.circle.fill",
            iconColor: .orange
        ))

        for topic in HelpTopic.allCases {
            items.append(SearchableItem(
                title: topic.title,
                subtitle: String(topic.helpText().prefix(100)),
                keywords: [topic.title.lowercased(), "help", "guide", "tutorial"],
                type: .helpTopic,
                destination: .specificHelpTopic(topic),
                icon: topic.icon,
                iconColor: .orange
            ))
        }

        // Privacy & Security
        items.append(SearchableItem(
            title: LocalizedString.searchPrivacyProtection.localized,
            subtitle: LocalizedString.searchDataProtected.localized,
            keywords: ["privacy", "security", "protection", "safe", "secure", "data", "zero knowledge", "zkp"],
            type: .helpTopic,
            destination: .specificHelpTopic(.zeroKnowledge),
            icon: "lock.shield",
            iconColor: .green
        ))

        // Verification
        items.append(SearchableItem(
            title: LocalizedString.searchAgeVerification.localized,
            subtitle: LocalizedString.searchProveAge.localized,
            keywords: ["verify", "verification", "age", "proof", "prove", "check"],
            type: .helpTopic,
            destination: .specificHelpTopic(.ageVerification),
            icon: "checkmark.shield",
            iconColor: .blue
        ))

        return items
    }
}

File: android/app/src/main/java/app/provii/wallet/search/SearchManager.kt
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust
package app.provii.wallet.search

import android.content.Context
import app.provii.wallet.utils.LanguageConfig
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Help
import androidx.compose.material.icons.filled.*
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import app.provii.wallet.R

enum class SearchableItemType {
    SCREEN,
    SETTING,
    HELP_TOPIC,
    FEATURE,
}

data class SearchableItem(
    val id: String,
    val title: String,
    val subtitle: String?,
    val keywords: List<String>,
    val type: SearchableItemType,
    val destination: SearchDestination,
    val icon: ImageVector,
    val iconColor: Color,
) {
    fun matches(query: String): Boolean {
        val lowercasedQuery = query.lowercase()
        return title.lowercase().contains(lowercasedQuery) ||
            (subtitle?.lowercase()?.contains(lowercasedQuery) ?: false) ||
            keywords.any { it.lowercase().contains(lowercasedQuery) }
    }
}

sealed class SearchDestination {
    object AccessibilitySettings : SearchDestination()

    object Settings : SearchDestination()

    object Credentials : SearchDestination()

    object Help : SearchDestination()

    object WhereToGet : SearchDestination()

    object LanguageSelection : SearchDestination()

    data class SpecificHelpTopic(val topicId: Int) : SearchDestination()
}

/**
 * In-app search engine that indexes screens, settings, help topics, and features
 * into a queryable catalogue. Search input is capped at 200 characters to match
 * the iOS implementation. Results are ranked by relevance: exact title match first,
 * then prefix match, then substring match, then alphabetical order.
 */
class SearchManager(context: Context) {
    var searchQuery by mutableStateOf("")
    var searchResults by mutableStateOf<List<SearchableItem>>(emptyList())
    var isSearching by mutableStateOf(false)

    private val allSearchableItems: List<SearchableItem> = createSearchableItems(context)

    fun search(query: String) {
        // Limit search input to 200 characters (matching iOS)
        searchQuery = query.take(200)

        if (searchQuery.isEmpty()) {
            searchResults = emptyList()
            isSearching = false
            return
        }

        isSearching = true

        // Filter items that match the (limited) query
        val matches = allSearchableItems.filter { it.matches(searchQuery) }

        // Sort by relevance
        searchResults =
            matches.sortedWith(
                compareBy(
                    // Exact matches first
                    { !it.title.equals(searchQuery, ignoreCase = true) },
                    // Title starts with query
                    { !it.title.startsWith(searchQuery, ignoreCase = true) },
                    // Title contains query
                    { !it.title.contains(searchQuery, ignoreCase = true) },
                    // Alphabetical
                    { it.title },
                ),
            )
    }

    fun clearSearch() {
        searchQuery = ""
        searchResults = emptyList()
        isSearching = false
    }

    companion object {
        private fun createSearchableItems(context: Context): List<SearchableItem> {
            val items = mutableListOf<SearchableItem>()

            // Accessibility Settings (HIGH PRIORITY)
            items.add(
                SearchableItem(
                    id = "accessibility_settings",
                    title = context.getString(R.string.search_item_accessibility_settings_title),
                    subtitle = context.getString(R.string.search_item_accessibility_settings_subtitle),
                    keywords =
                        listOf(
                            "accessibility", "a11y", "settings", "customize", "screen reader",
                            "voiceover", "talkback", "large text", "high contrast", "voice",
                            "speech", "disabilities", "wcag", "ada",
                        ),
                    type = SearchableItemType.SCREEN,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.AccessibilityNew,
                    iconColor = Color.Blue,
                ),
            )

            // Individual Accessibility Features
            items.add(
                SearchableItem(
                    id = "large_text",
                    title = context.getString(R.string.search_item_large_text_title),
                    subtitle = context.getString(R.string.search_item_large_text_subtitle),
                    keywords = listOf("large", "text", "size", "font", "bigger", "accessibility", "vision", "sight"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.TextFields,
                    iconColor = Color.Blue,
                ),
            )

            items.add(
                SearchableItem(
                    id = "high_contrast",
                    title = context.getString(R.string.search_item_high_contrast_title),
                    subtitle = context.getString(R.string.search_item_high_contrast_subtitle),
                    keywords = listOf("high", "contrast", "visibility", "vision", "accessibility", "colors", "see better"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.Contrast,
                    iconColor = Color.Blue,
                ),
            )

            items.add(
                SearchableItem(
                    id = "voice_input",
                    title = context.getString(R.string.search_item_voice_input_title),
                    subtitle = context.getString(R.string.search_item_voice_input_subtitle),
                    keywords = listOf("voice", "speech", "input", "control", "speak", "talk", "accessibility", "hands-free"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.Mic,
                    iconColor = Color.Blue,
                ),
            )

            items.add(
                SearchableItem(
                    id = "manual_code_entry",
                    title = context.getString(R.string.search_item_manual_code_entry_title),
                    subtitle = context.getString(R.string.search_item_manual_code_entry_subtitle),
                    keywords = listOf("manual", "code", "entry", "type", "keyboard", "accessibility", "qr", "scan"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.Keyboard,
                    iconColor = Color.Blue,
                ),
            )

            items.add(
                SearchableItem(
                    id = "simplified_ui",
                    title = context.getString(R.string.search_item_simplified_ui_title),
                    subtitle = context.getString(R.string.search_item_simplified_ui_subtitle),
                    keywords = listOf("simplified", "simple", "ui", "interface", "easy", "basic", "accessibility", "cognitive"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.GridView,
                    iconColor = Color.Blue,
                ),
            )

            items.add(
                SearchableItem(
                    id = "color_blindness",
                    title = context.getString(R.string.search_item_color_blindness_title),
                    subtitle = context.getString(R.string.search_item_color_blindness_subtitle),
                    keywords = listOf("color", "blind", "blindness", "vision", "accessibility", "deuteranopia", "protanopia", "tritanopia"),
                    type = SearchableItemType.SETTING,
                    destination = SearchDestination.AccessibilitySettings,
                    icon = Icons.Default.Palette,
                    iconColor = Color.Blue,
                ),
            )

            // Settings
            items.add(
                SearchableItem(
                    id = "settings",
                    title = context.getString(R.string.search_item_settings_title),
                    subtitle = context.getString(R.string.search_item_settings_subtitle),
                    keywords = listOf("settings", "config", "configuration", "preferences", "options"),
                    type = SearchableItemType.SCREEN,
                    destination = SearchDestination.Settings,
                    icon = Icons.Default.Settings,
                    iconColor = Color.Gray,
                ),
            )

            if (LanguageConfig.hasMultipleLanguages) {
                items.add(
                    SearchableItem(
                        id = "language",
                        title = context.getString(R.string.search_item_language_title),
                        subtitle = context.getString(R.string.search_item_language_subtitle),
                        keywords = listOf("language", "translate", "locale", "español", "français", "deutsch", "italiano", "português"),
                        type = SearchableItemType.SCREEN,
                        destination = SearchDestination.LanguageSelection,
                        icon = Icons.Default.Language,
                        iconColor = Color.Green,
                    ),
                )
            }

            // Credentials
            items.add(
                SearchableItem(
                    id = "credentials",
                    title = context.getString(R.string.search_item_credentials_title),
                    subtitle = context.getString(R.string.search_item_credentials_subtitle),
                    keywords = listOf("credentials", "wallet", "id", "verification", "identity"),
                    type = SearchableItemType.SCREEN,
                    destination = SearchDestination.Credentials,
                    icon = Icons.Default.Wallet,
                    iconColor = Color(0xFF9C27B0), // Purple
                ),
            )

            items.add(
                SearchableItem(
                    id = "get_credential",
                    title = context.getString(R.string.search_item_get_credential_title),
                    subtitle = context.getString(R.string.search_item_get_credential_subtitle),
                    keywords = listOf("get", "obtain", "acquire", "credential", "issuer", "location", "where"),
                    type = SearchableItemType.SCREEN,
                    destination = SearchDestination.WhereToGet,
                    icon = Icons.Default.Add,
                    iconColor = Color.Blue,
                ),
            )

            // Help Topics
            items.add(
                SearchableItem(
                    id = "help",
                    title = context.getString(R.string.search_item_help_title),
                    subtitle = context.getString(R.string.search_item_help_subtitle),
                    keywords = listOf("help", "support", "assistance", "guide", "tutorial", "how to"),
                    type = SearchableItemType.SCREEN,
                    destination = SearchDestination.Help,
                    icon = Icons.AutoMirrored.Filled.Help,
                    iconColor = Color(0xFFFF9800), // Orange
                ),
            )

            // Privacy & Security
            items.add(
                SearchableItem(
                    id = "privacy",
                    title = context.getString(R.string.search_item_privacy_title),
                    subtitle = context.getString(R.string.search_item_privacy_subtitle),
                    keywords = listOf("privacy", "security", "protection", "safe", "secure", "data", "zero knowledge", "zkp"),
                    type = SearchableItemType.HELP_TOPIC,
                    destination = SearchDestination.SpecificHelpTopic(4),
                    icon = Icons.Default.Security,
                    iconColor = Color.Green,
                ),
            )

            // Verification
            items.add(
                SearchableItem(
                    id = "verification",
                    title = context.getString(R.string.search_item_verification_title),
                    subtitle = context.getString(R.string.search_item_verification_subtitle),
                    keywords = listOf("verify", "verification", "age", "proof", "prove", "check"),
                    type = SearchableItemType.HELP_TOPIC,
                    destination = SearchDestination.SpecificHelpTopic(2),
                    icon = Icons.Default.VerifiedUser,
                    iconColor = Color.Blue,
                ),
            )

            return items
        }
    }
}
